### ISSUE TITLE
Feature request: support floating-point numbers

### ISSUE THREAD

**Issue by justinlovinger** [HdrHistogram/HdrHistogram_rust#128]
# Feature request: support floating-point numbers

See <https://github.com/HdrHistogram/HdrHistogram_rust/issues/91>,

> Floating point values _can_ be supported, but they are basically a completely [separate implementation](https://github.com/HdrHistogram/HdrHistogram/blob/master/src/main/java/org/HdrHistogram/DoubleHistogram.java) that wraps an inner integer-based histogram. I'd be neat to port `DoubleHistogram` to Rust as well, but unfortunately it's a larger effort that I don't have time for myself at the moment. If you want to give it a try, I'd be happy to try and review? It _should_ be a fairly straightforward translation of the Java version I linked above! I wonder if we may even be able to have a _single_ `Histogram` type by providing inherent implementations for `Histogram<f64>`... Not entirely clear. Initially I'd just have a separate `DoubleHistogram` type.

If possible, I suggest implementing a type-generic histogram using `num-traits`.

**Comment by jonhoo** [comment]
It won't be possible to use num-trait here I don't think. Floats have to be handled fundamentally differently in HDR histograms is my understanding.

**Comment by justinlovinger** [comment]
But could all floats be handled with the same implementation? I meant something like `RealHistogram<A: Real>` or `FloatHistogram<A: Float>`.

Unfortunately, we cannot do the same for integer histograms because `num-traits` lacks an "integer but not real" trait to differentiate numbers with whole values from numbers with whole values *and* decimals.

**Comment by jonhoo** [comment]
Probably, though we could do that with a trait of our own that we implement just for `f32` and `f64` rather than going via `num-traits`. That way we're also reducing the breaking change hazard. Alternatively _just_ implement for `f64`.

**Comment by justinlovinger** [comment]
If we implement our own trait, it will not work with other types implementing `num-traits` traits. For example, the application I want to use HdrHistogram for has a type wrapping `f64`. `num-traits` is already used extensively in the Rust ecosystem.

**Comment by jonhoo** [comment]
Yup, well aware of the implications. I'd want to see how the implementation ultimately pans out — I suspect going via `num-traits` might end up quite painful (vs just supporting `f64`), but happy to be proven wrong.
